index.d.ts

```diff
@@ -4,6 +4,12 @@ export interface FastifyError extends Error {
   validation?: ValidationResult[]
 }
 
+export interface FastifyErrorConstructor extends ErrorConstructor {
```

Comment:
Are we sure this is the right pattern to use here? I think FastifyError should be defined as a class rather than an interface and then a lot of this is implicit.

Reply:
I also tried the elasticsearch style but the tests fail "This expression is not constructable  Type 'FastifyErrorConstructor' has no construct signatures"
![Screenshot 2020-10-09 at 10 41 28](https://user-images.githubusercontent.com/1814734/95556456-fffb8300-0a1b-11eb-9a92-f34eb013aae6.png)

Style I used is copied from the typescript's own definitions
![Screenshot 2020-10-09 at 9 58 37](https://user-images.githubusercontent.com/1814734/95553117-8f9e3300-0a16-11eb-9df5-8e88f0e8cb5d.png)
.

Reply:
if you have error defined as a class you shouldn't need to define a separate constructor

Reply:
modify line 1 from an interface to class (if possible). remove this FastifyErrorConstructor business, and then you'll need to do something with the return type on createError.

Reply:
@Ethan-Arrowood could you please try it yourself? I have tried the class method, with and without class constructor and also implements/extends combinations but all of those result in  "index.test-d.ts:8:0
  ✖   7:18  This expression is not constructable.
  Type FastifyError has no construct signatures."

It might be a bug in the "tsd" used in the project or something to do with typescript 4 as the error is defined internally as a interface.

Reply:
Ahh good point, I'll take a look later and see what's up. I'm only confused because I'm pretty sure I did something similar to this for the experimental undici types and used classes. Thanks for trying again, I'll post an update later today if I have the opportunity!

Reply:
I wrote the module directly in typescript to find out what it outputs. The Base in createError changed to ErrorConstructor and FastifyError got constructor method. Howerver what is not possible with typescript classes is the real implementations static constructor (creating instance without new). So my initial typedefinition would be most true to real implementation.
![Screenshot 2020-10-10 at 13 52 56](https://user-images.githubusercontent.com/1814734/95653472-b2594600-0b01-11eb-952c-78aa92172d42.png)

Reply:
Today I learned! Thank you for this. I'll go ahead and approve now :)

Reply:
Just out of curiosity can you try using `InstanceType<FastifyError>` (here the [reference](https://github.com/Microsoft/TypeScript/blob/2961bc3fc0ea1117d4e53bc8e97fa76119bc33e3/src/lib/es5.d.ts#L1466-L1469)) instead of `typeof FastifyError` in the experiment you did? It should work without problems, and it is more correct: I really don't like that `readonly prototype:`.

Reply:
@fox1t did try but didn't quite get how would in this situation as we can't have static constructors in ts-classes https://basarat.gitbook.io/typescript/main-1/staticconstructor